umbra-js/classes/KeyPair.js

```diff
@@ -27,17 +28,60 @@ function pad32ByteHex(hex) {
   return hex.padStart(64, 0);
 }
 
+/**
+ * @notice Given a transaction hash, return the public key of the transaction's sender
+ * @dev See https://github.com/ethers-io/ethers.js/issues/700 for an example of
+ * recovering public key from a transaction with ethers
+ * @param {String} txHash Transaction hash to recover public key from
+ */
+async function recoverPublicKeyFromTransaction(txHash) {
+  // Get transaction data
+  const tx = await provider.getTransaction(txHash);
+
+  // Get original signature
+  const splitSignature = {
+    r: tx.r,
+    s: tx.s,
+    v: tx.v,
+  };
+  const signature = ethers.utils.joinSignature(splitSignature);
+
+  // Reconstruct transaction data that was originally signed
+  const txData = {
+    chainId: tx.chainId,
+    data: tx.data,
+    gasLimit: tx.gasLimit,
+    gasPrice: tx.gasPrice,
+    nonce: tx.nonce,
+    to: tx.to, // this works for both regular and contract transactions
+    value: tx.value,
+  };
+
+  // Properly format it to get the correct message
+  const resolvedTx = await ethers.utils.resolveProperties(txData);
+  const rawTx = ethers.utils.serializeTransaction(resolvedTx); // returns RLP encoded tx
+  const msgHash = ethers.utils.keccak256(rawTx); // as specified by ECDSA
+  const msgBytes = ethers.utils.arrayify(msgHash); // create binary hash
+
+  // Recover sender's public key and address
+  const publicKey = ethers.utils.recoverPublicKey(msgBytes, signature);
+  return publicKey;
+}
+
 class KeyPair {
   /**
-   * @notice Creates new instance from a public or private key
-   * @param {String} key public or private key to initialize instance with, as hex with prefix
+   * @notice Creates new instance from a public key, private key, or transaction hash
+   * @param {String} key Can be either (1) hex public key with 0x04 prefix, (2) hex private
+   * key with 0x prefix, or (3) transaction hash
+   * @param {Boolean} isTxHash must be true if key is a transaction hash
    */
-  constructor(key) {
+  constructor(key, isTxHash = false) {
```

Comment:
Generally I consider boolean flags being passed to functions to be a code smell. Maybe combined with extracting the async action, as mentioned in my other comment, we can avoid requiring this?

Reply:
Agree with this, but can probably avoid it using whatever solution we decide upon in the other comment